lib/matplotlib/pyplot.py

```diff
@@ -200,6 +200,126 @@ def uninstall_repl_displayhook():
 
 draw_all = _pylab_helpers.Gcf.draw_all
 
+_ENSURE_AX_DOC = """
+This function has been decorated by pyplot to have
+an implicit reference to the `plt.gca()` passed as the first argument.
+
+The wrapped function can be called as any of ::
+
+   func(*args, **kwargs)
+   func(ax, *args, **kwargs)
+   func(.., ax=ax)
+
+"""
+
+
+def ensure_ax(func):
+    """Decorator to ensure that the function gets an `Axes` object.
+
+
+    The intent of this decorator is to simplify the writing of helper
+    plotting functions that are useful for both interactive and
+    programmatic usage.
+
+    The encouraged signature for third-party and user functions ::
+
+       def my_function(ax, data, style)
+
+    explicitly expects an Axes object as input rather than using
+    plt.gca() or creating axes with in the function body.  This
+    allows for great flexibility, but some find it verbose for
+    interactive use.  This decorator allows the Axes input to be
+    omitted in which case `plt.gca()` is passed into the function.
+    Thus ::
+
+       wrapped = ensure_ax(my_function)
+
+    can be called as any of ::
+
+       wrapped(data, style)
+       wrapped(ax, data, style)
+       wrapped(data, style, ax=plt.gca())
+
+
+    """
+    @wraps(func)
+    def inner(*args, **kwargs):
+        if 'ax' in kwargs:
+            ax = kwargs.pop('ax', None)
+        elif len(args) > 0 and isinstance(args[0], Axes):
+            ax = args[0]
+            args = args[1:]
+        else:
+            ax = gca()
+        return func(ax, *args, **kwargs)
+    pre_doc = inner.__doc__
+    if pre_doc is None:
+        pre_doc = ''
+    else:
+        pre_doc = dedent(pre_doc)
+    inner.__doc__ = _ENSURE_AX_DOC + pre_doc
+
+    return inner
+
+
+def ensure_ax_meth(func):
+    """
+    The same as ensure axes, but for class methods ::
+
+       class foo(object):
+           @ensure_ax_meth
+           def my_function(self, ax, style):
+
+    will allow you to call your objects plotting methods with
+    out explicitly passing in an `Axes` object.
+    """
+    @wraps(func)
+    def inner(*args, **kwargs):
+        s = args[0]
+        args = args[1:]
+        if 'ax' in kwargs:
+            ax = kwargs.pop('ax', None)
+        elif len(args) > 1 and isinstance(args[0], Axes):
+            ax = args[0]
+            args = args[1:]
+        else:
+            ax = gca()
+        return func(s, ax, *args, **kwargs)
+    pre_doc = inner.__doc__
+    if pre_doc is None:
+        pre_doc = ''
+    else:
+        pre_doc = dedent(pre_doc)
+    inner.__doc__ = _ENSURE_AX_DOC + pre_doc
+    return inner
+
+
+def register_func(func):
+    """
+    Registers a function to be wrapped and made available in the
+    pyplot namespace for convenient interactive command line use.
```

Comment:
I want to keep something in there so we don't encourage people to use these
functions in other functions.

I am a bit hesitant to include this at all for exactly that reason.

On Tue, Jul 7, 2015, 3:06 PM Benjamin Root <notifications@github.com> wrote:

> In lib/matplotlib/pyplot.py
> <https://github.com/matplotlib/matplotlib/pull/4488#discussion_r34083841>:
>
> > +        else:
> > +            ax = gca()
> > +        return func(s, ax, *args, **kwargs)
> > +    pre_doc = inner.__doc__
> > +    if pre_doc is None:
> > +        pre_doc = ''
> > +    else:
> > +        pre_doc = dedent(pre_doc)
> > +    inner.__doc__ = _ENSURE_AX_DOC + pre_doc
> > +    return inner
> > +
> > +
> > +def register_func(func):
> > +    """
> > +    Registers a function to be wrapped and made available in the
> > +    pyplot namespace for convenient interactive command line use.
>
> command-line use? Might be misleading. Would probably be better as just
> "interactive use".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4488/files#r34083841>.
>